diff --git a/Makefile b/Makefile
new file mode 100644
index 00000000..e89ec27b
--- /dev/null
+++ b/Makefile
@@ -0,0 +1,16 @@
+
+
+all: setup build install
+
+setup:
+	CC=clang \
+	CXX=clang++ \
+	meson setup -Dprefix=/usr --wipe build
+
+build: setup
+	CC=clang \
+	CXX=clang++ \
+	ninja -C build
+
+install: build
+	sudo ninja -C build install
diff --git a/gthumb/gth-time-selector.c b/gthumb/gth-time-selector.c
index 270c6946..9542fec3 100644
--- a/gthumb/gth-time-selector.c
+++ b/gthumb/gth-time-selector.c
@@ -20,7 +20,11 @@
  */
 
 #include <config.h>
+#if defined(__QNX__)
+#define _XOPEN_SOURCE 700 /* glibc2 needs this */
+#else
 #define _XOPEN_SOURCE /* glibc2 needs this */
+#endif
 #define _XOPEN_SOURCE_EXTENDED 1
 #include <time.h>
 #include <glib/gi18n.h>
diff --git a/meson.build b/meson.build
index f7da6251..e22bf9db 100644
--- a/meson.build
+++ b/meson.build
@@ -239,7 +239,7 @@ common_deps = [
   dependency('gmodule-2.0'),
   dependency('gio-unix-2.0'),
   dependency('gtk+-3.0', version : gtk_version),
-  dependency('x11'),
+#  dependency('x11'),
   dependency('libpng'),
   dependency('zlib'),
   dependency('gsettings-desktop-schemas')
