--- ../../build/qemu_qsti/output/build/mount_fs.sh.orig	2025-09-25 10:54:28.605065241 -0400
+++ ../../build/qemu_qsti/output/build/mount_fs.sh	2025-09-25 10:55:28.817338177 -0400
@@ -16,7 +16,7 @@
 
 echo "---> Mounting file systems"
 
-if [ -e /data -a -e /system ]; then
+if [ -e /data -a -e /usr ]; then
 	exit 0
 fi
 
@@ -33,22 +33,22 @@
 	# container and then the file system contained within
 	mount -t qtd -o key=/proc/boot/qtd_public_key.pem  /dev/hd0t185 /dev/hd0t185-fs
 	if check_magic /dev/hd0t185-fs 8192 "\0042\0021\0031\0150"; then
-		mount -t qnx6 -o noatime  /dev/hd0t185-fs /system
+		mount -t qnx6 -o noatime  /dev/hd0t185-fs /usr
 	elif check_magic /dev/hd0t185-fs 0 "\0377\0272\0377\0261"; then
-		mount -t qcfs  /dev/hd0t185-fs /system
+		mount -t qcfs  /dev/hd0t185-fs /usr
 	else
 		echo "Don't recognise the filesystem"
 	fi
 	rm -f /dev/shmem/fs-magic*
 elif [ -e /dev/hd0t186 ]; then
-    qtsafefsd -o key=/proc/boot/qtsafefs_public_key.pem  /dev/hd0t186 /system
+    qtsafefsd -o key=/proc/boot/qtsafefs_public_key.pem  /dev/hd0t186 /usr
 elif [ -e /dev/hd0t181 ]; then
-	mount -t qcfs  /dev/hd0t181 /system
+	mount -t qcfs  /dev/hd0t181 /usr
 else # QNX6
-	mount -t qnx6 -o noatime  /dev/hd0t178 /system 
+	mount -t qnx6 -o noatime  /dev/hd0t178 /usr 
 fi
 
 
 mount -t qnx6  /dev/hd0t179 /data 
 
-/system/bin/cleanup_tmp
+/usr/bin/cleanup_tmp
